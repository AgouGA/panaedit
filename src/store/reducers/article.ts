/* add article to project */
export function _addArticle(state: State, action: AddArticleAction): State {
	const { article } = action.payload

	state.articles.push(article)

	return state
}

/* remove an article from the project */
export function _removeArticle(state: State, action: RemoveArticleAction): State {
	const { articleId } = action.payload

	// find the article by id
	const index = state.articles.findIndex((article) => article.id == articleId)

	if (index == -1) {
		throw new Error('cannot delete, article does not exist')
	}

	const removed = state.articles.splice(index, 1)

	// TODO: remove all references to this article from the project

	return state
}
