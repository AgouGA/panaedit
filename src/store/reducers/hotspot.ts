import _ from 'lodash'

/* add hotspot */
export function _addHotspot(state: State, action: AddHotspotAction) {
	const { sceneKey, hotspot } = action.payload
	const scene = state.scenes[sceneKey]

	scene.hotSpots.push(hotspot)

	return scene
}

/* remove hotspot */
export function _removeHotspot(state: State, action: RemoveHotspotAction) {
	const { sceneKey, hotspotIndex } = action.payload
	const scene = state.scenes[sceneKey]

	_.pullAt(scene.hotSpots, hotspotIndex)

	return scene
}
