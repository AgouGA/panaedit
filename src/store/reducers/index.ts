/*
 *
 */

export * from './article'
export * from './editor'
export * from './hotspot'
// index.ts
export * from './photo'
export * from './project'
export * from './scene'
