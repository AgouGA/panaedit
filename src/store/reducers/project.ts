import { AnyAction, Action } from 'redux'
import stateDefaultState from '../storeDefaultState'
import _ from 'lodash'

/* restore the project to defaults */
export function _reset(state: State) {
	return stateDefaultState
}

/* load project json from disc. Should validate first tho */
export function _loadProject(state: State, action: AnyAction) {
	return action.payload.project
}
