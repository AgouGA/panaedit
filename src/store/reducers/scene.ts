import _ from 'lodash'

/* add scene. we use the filename as the key in the scene array */
export function _addScene(state: State, action: AddSceneAction) {
	const { panorama } = action.payload.scene

	if (!panorama) {
		throw new Error('scene has no panorama file path')
	}

	// find the file name from the path, remove file extension
	// having '.' in the key makes lodash create more nested objects
	const fname = _.last(panorama.split('/'))
	const key = _.first(fname?.split('.'))

	if (!key) {
		throw new Error('invalid panorama name / scene key')
	}

	if (state.scenes[key]) {
		throw new Error('scene already exists')
	}

	// add scene to state,
	_.set(state.scenes, key, action.payload.scene)

	return state
}

/* remove scene */
export function _removeScene(state: State, action: RemoveSceneAction) {
	const { sceneKey } = action.payload

	if (!state.scenes[sceneKey]) {
		throw new Error('scene does not exist')
	}

	// remove the scene
	state.scenes = _.omit(state.scenes, sceneKey)

	return state
}
