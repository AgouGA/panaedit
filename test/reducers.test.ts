import { resetReducer } from '../src/store/reducers/scene'
import { describe, it, expect } from 'vitest'

it('should return the default', () => {
	expect(resetReducer).toBeDefined()

	const input = {
		hello: 'there'
	}

	const output = resetReducer(input as any, { type: 'reset' })

	expect(output).toEqual(input)
})
