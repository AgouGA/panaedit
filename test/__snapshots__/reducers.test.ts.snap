// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test reducer functions > _reset should return the default project 1`] = `
{
  "articles": [],
  "default": {
    "autoLoad": true,
    "basePath": "assets/panorama/",
    "compass": true,
    "firstScene": "",
    "hfov": 110,
    "hotSpotDebug": true,
    "imagePath": "assets/images/",
    "maxPitch": 98,
    "minPitch": -98,
    "sceneFadeDuration": 1000,
    "type": "equirectangular",
    "vfov": 100,
  },
  "editor": {
    "currentScene": "",
  },
  "scenes": {},
}
`;

exports[`test reducer functions > add scene should add new key to scenes obj 1`] = `
{
  "articles": [],
  "default": {
    "autoLoad": true,
    "basePath": "assets/panorama/",
    "compass": true,
    "firstScene": "",
    "hfov": 110,
    "hotSpotDebug": true,
    "imagePath": "assets/images/",
    "maxPitch": 98,
    "minPitch": -98,
    "sceneFadeDuration": 1000,
    "type": "equirectangular",
    "vfov": 100,
  },
  "editor": {
    "currentScene": "",
  },
  "scenes": {
    "testimage": {
      "hotSpots": [],
      "id": "testid",
      "name": "test object",
      "panorama": "./assets/testimage.jpg",
      "type": "link",
    },
  },
}
`;
